src/ol_infrastructure/lib/vault.py

```diff
@@ -17,134 +18,292 @@
+            ),
+            Template(
+                """REVOKE ALL PRIVILEGES ON ALL SEQUENCES IN SCHEMA public FROM "{{name}}";"""
+            ),
+            Template("""REVOKE USAGE ON SCHEMA public FROM "{{name}}";"""),
+            # Finally, drop this user from the database
+            Template("""DROP USER "{{name}}";"""),
+        ],
+        "renew": [],
+        "rollback": [],
     },
     "app": {
-        "create": Template(
-            """CREATE USER "{{name}}" WITH PASSWORD '{{password}}'
-            VALID UNTIL '{{expiration}}' IN ROLE "${app_name}" INHERIT;
-          GRANT CREATE ON SCHEMA public TO ${app_name} WITH GRANT OPTION;
-          GRANT ALL PRIVILEGES ON ALL TABLES IN SCHEMA public TO "${app_name}"
-             WITH GRANT OPTION;
-          GRANT ALL PRIVILEGES ON ALL SEQUENCES IN SCHEMA public TO "${app_name}"
-             WITH GRANT OPTION;
-          SET ROLE "${app_name}";
-          ALTER DEFAULT PRIVILEGES FOR ROLE "${app_name}" IN SCHEMA public
-            GRANT ALL PRIVILEGES ON TABLES TO "${app_name}" WITH GRANT OPTION;
-          ALTER DEFAULT PRIVILEGES FOR ROLE "${app_name}" IN SCHEMA public
-            GRANT ALL PRIVILEGES ON SEQUENCES TO "${app_name}" WITH GRANT OPTION;
-          RESET ROLE;
-          ALTER ROLE "{{name}}" SET ROLE "${app_name}";"""
-        ),
-        "revoke": Template(
-            """REVOKE "${app_name}" FROM "{{name}}";
-          GRANT "{{name}}" TO ${app_name} WITH ADMIN OPTION;
-          SET ROLE ${app_name};
-          REASSIGN OWNED BY "{{name}}" TO "${app_name}";
-          RESET ROLE;
-          DROP OWNED BY "{{name}}";
-          REVOKE ALL PRIVILEGES ON ALL TABLES IN SCHEMA public FROM "{{name}}";
-          REVOKE ALL PRIVILEGES ON ALL SEQUENCES IN SCHEMA public FROM "{{name}}";
-          REVOKE USAGE ON SCHEMA public FROM "{{name}}";
-          DROP USER "{{name}}";"""
-        ),
+        "create": [
+            # Check if the role exists and create it if not
+            Template(
+                """
+                DO
+                $$do$$
+                BEGIN
+                   IF EXISTS (
+                      SELECT FROM pg_catalog.pg_roles
+                      WHERE  rolname = '${app_name}') THEN
+                   ELSE
+                      BEGIN   -- nested block
+                         CREATE ROLE ${app_name};
+                      EXCEPTION
+                         WHEN duplicate_object THEN
+                            RAISE NOTICE 'Role "${app_name}" was just created by a concurrent transaction. Skipping.';
+                      END;
+                   END IF;
+                END
+                $$do$$;
+                """
+            ),
+            # Set/refresh the default privileges for the new role
+            Template(
+                """GRANT CREATE ON SCHEMA public TO ${app_name} WITH GRANT OPTION;"""
+            ),
+            Template("""
```

Comment:
Do we still want to apply all of these grants during the user creation step? I'm just curious about the benefit of not having the approle definition for assigning/updating the role permissions and then just having the app/readonly/etc. just get assigned the appropriate role?

Reply:
So I think the primary benifit here is that the role gets 'fixed' every time a new user is created and assigned too it. If there were any out-of-band changes that took place to the role, at least this minimum set of privs will be restored. It won't address any _additional_ permissions that may have been given but it will at least restore the minimum set.

We can move this into the fancy conditional structure when the role is now 'created' or its existence is checked but it will have the same effect. If we put these steps up there, each line will have to exist twice in both steps of the conditional. 

I believe moving these grants / default privs into their own statements like this will address error message we get about block size on new account provisions.